# EJERCICIOS DEL LIBRO

De acuerdo al libro Quantum Computing For Computer Scientists. Se realizaran los ejercicios de programación del capitulo 1 con el lenguaje python.

## CAPITULO 1

### IMPRIMIR COMPLEJOS

In [2]:
def imprimir(Cm1):
    Cm1 = Cm1[:]
    
    if Cm1[0]==0:
        print("Numero complejo :",str(Cm1[1])+"i")
    elif Cm1[1]==0:
        print("Numero complejo :",Cm1[0])
    elif Cm1[1]<0:
        print("Numero complejo :",Cm1[0],str(Cm1[1])+"i")
    else:
        print("Numero complejo :",Cm1[0],"+"+str(Cm1[1])+"i")
    
    

### 1.1 SUMA DE COMPLEJOS

Programa que sume dos numeros complejos.

In [3]:
import math

In [4]:
def suma_C(C1,C2):
    C1 = C1[:]
    C2 = C2[:]
    sum_com = ((C1[0]+C2[0]),(C1[1]+C2[1]))
    print("----------SUMA DE DOS NÚMEROS-------")
    #Numeros
    imprimir(C1)
    imprimir(C2)
    #Resultado
    print("----------RESULTADO-------")
    imprimir(sum_com)
    
     #Si alguna otra función necesita multiplicar dos numeros se devuelve el resultado
    return sum_com

In [5]:
suma_C((3,5),(0,2))

----------SUMA DE DOS NÚMEROS-------
Numero complejo : 3 +5i
Numero complejo : 2i
----------RESULTADO-------
Numero complejo : 3 +7i


(3, 7)

#### PRUEBA
Se realizo las siguientes pruebas con dos numero complejos

Numero 1: $$ 3 +5i $$
Numero 2: $$ 2i $$
Y se obtuvo como resultado el siguiente numero
$$ 3 +7i $$

### 1.2 RESTA DE COMPLEJOS

In [6]:
def resta_C(C1,C2):
    C1 = C1[:]
    C2 = C2[:]
    resta = ((C1[0]-C2[0]),(C1[1]-C2[1]))
    print("----------SUMA DE DOS NÚMEROS-------")
    #Numeros
    imprimir(C1)
    imprimir(C2)
    #Resultado
    print("----------RESULTADO-------")
    imprimir(resta)
    
     #Si alguna otra función necesita multiplicar dos numeros se devuelve el resultado
    return resta

In [7]:
resta_C((3,5),(0,2))

----------SUMA DE DOS NÚMEROS-------
Numero complejo : 3 +5i
Numero complejo : 2i
----------RESULTADO-------
Numero complejo : 3 +3i


(3, 3)

#### PRUEBA
Se realizo las siguientes pruebas con dos numero complejos

Numero 1: $$ 3 +5i $$
Numero 2: $$ 2i $$
Y se obtuvo como resultado el siguiente numero
$$ 3 +3i $$

### 1.3 MULTIPLICACION DE COMPLEJOS

Programa que multiplique dos numeros complejos

In [8]:
def Mul_C(C1,C2):
    C1 = C1[:]
    C2 = C2[:]
    Real = (C1[0]*C2[0])-(C1[1]*C2[1])
    imag = (C1[0]*C2[1])+(C2[0]*C1[1])
    Resul = (Real,imag)
    print("----------MULTIPLICACIÓN DE DOS NÚMEROS-------")
    #Numeros
    imprimir(C1)
    imprimir(C2)
    #Resultado
    print("----------RESULTADO-------")
    imprimir(Resul)
    
    
    #Si alguna otra función necesita multiplicar dos numeros se devuelve el resultado
    return Resul

In [9]:
Mul_C((-1,0),(1,3))

----------MULTIPLICACIÓN DE DOS NÚMEROS-------
Numero complejo : -1
Numero complejo : 1 +3i
----------RESULTADO-------
Numero complejo : -1 -3i


(-1, -3)

#### PRUEBA
Se realizo las siguientes pruebas con dos numero complejos

Numero 1: $$ -1 $$
Numero 2: $$ 1 +3i $$
Y se obtuvo como resultado el siguiente numero
$$ -1 -3i $$

### 1.4 CONJUGADO DE UN NUMERO COMPLEJO

Programa que conjugue un numero complejo

In [10]:
def Conju_C(C1):
    C1 = C1[:]
    #Conjugado
    Resul = (C1[0],-C1[1])
    
    print("----------CONJUGADO DE UN NÚMERO-------")
    #Numeros
    imprimir(C1)
    #Resultado
    print("----------RESULTADO-------")
    imprimir(Resul)
        
    return Resul

In [11]:
Conju_C((-5,4))

----------CONJUGADO DE UN NÚMERO-------
Numero complejo : -5 +4i
----------RESULTADO-------
Numero complejo : -5 -4i


(-5, -4)

#### PRUEBA
Se realizo las siguientes pruebas con un numero complejos

Numero : $$ -5 +4i $$
Y se obtuvo como resultado el siguiente numero
$$ -5 -4i $$

### 1.5 MODULO NUMERO COMPLEJO

Programa que muestre el modulo de un numero comlejo

In [12]:
def Modu_C(C1):
    C1 = C1[:]
    Resul = math.sqrt(C1[0]**2 + C1[1]**2)
    print("----------MODULO DE UN NÚMERO-------")
    #Numeros
    imprimir(C1)
    #Resultado
    print("----------RESULTADO-------")
    print(Resul)
    
    return Resul

In [13]:
Modu_C([-5, 4])

----------MODULO DE UN NÚMERO-------
Numero complejo : -5 +4i
----------RESULTADO-------
6.4031242374328485


6.4031242374328485

#### PRUEBA
Se realizo las siguientes pruebas con un numero complejos

Numero : $$ -5 +4i$$
Y se obtuvo como resultado el siguiente numero
$$ 6.4031242374328485 $$

### 1.6 DIVISIÓN NÚMEROS COMPLEJOS

Programa que Divida dos numeros complejos

In [14]:
def Div_C(C1, C2):
    
    if not(C2[0]==0 and C2[1]==0):
        Conjugado = Conju_C(C2)
        Numerador = Mul_C(C1, Conjugado)
        Denominador = Mul_C(C2, Conjugado)
        Resul = (Numerador,Denominador)
        print("----------DIVISION DE DOS NÚMEROS-------")
        #Numeros
        imprimir(C1)
        imprimir(C2)
        #Resultado
        print("----------RESULTADO-------")
        print("Numerador----> ")
        imprimir(Resul[0])
        print("Denominador----> ")
        imprimir(Resul[1])
        return Resul
    else:
        return False

In [15]:
Div_C((-5,4),(1,3))

----------CONJUGADO DE UN NÚMERO-------
Numero complejo : 1 +3i
----------RESULTADO-------
Numero complejo : 1 -3i
----------MULTIPLICACIÓN DE DOS NÚMEROS-------
Numero complejo : -5 +4i
Numero complejo : 1 -3i
----------RESULTADO-------
Numero complejo : 7 +19i
----------MULTIPLICACIÓN DE DOS NÚMEROS-------
Numero complejo : 1 +3i
Numero complejo : 1 -3i
----------RESULTADO-------
Numero complejo : 10
----------DIVISION DE DOS NÚMEROS-------
Numero complejo : -5 +4i
Numero complejo : 1 +3i
----------RESULTADO-------
Numerador----> 
Numero complejo : 7 +19i
Denominador----> 
Numero complejo : 10


((7, 19), (10, 0))

#### PRUEBA
Se realizo las siguientes pruebas con dos numero complejos

$$ -5 +4i  /  2i  $$
Y se obtuvo como resultado el siguiente numero
$$ 7 +19i  /  10 $$

### 1.7 FORMA RECTAGULAR A POLAR

Programa que muestre la forma polar de un numero rectangular dado

In [16]:
def Rec_Polar(C1):
    p = Modu_C(C1)
    Teta = math.atan2(C1[1],C1[0])
    Resul = [p, Teta]
    print("----------FORMA RECTANGULAR A POLAR-------")
    #Numeros
    imprimir(C1)
    #Resultado
    print("----------RESULTADO-------")
        
    print("Forma polar: (",Resul[0],",",Resul[1],")" )

In [17]:
Rec_Polar([-2, -1])

----------MODULO DE UN NÚMERO-------
Numero complejo : -2 -1i
----------RESULTADO-------
2.23606797749979
----------FORMA RECTANGULAR A POLAR-------
Numero complejo : -2 -1i
----------RESULTADO-------
Forma polar: ( 2.23606797749979 , -2.677945044588987 )


#### PRUEBA
Se realizo las siguientes pruebas con dos numero complejos

$$ -2 -1i  $$
Y se obtuvo como resultado el siguiente numero
$$ ( 2.23606797749979 , -2.677945044588987 ) $$

### 1.8 FORMA POLAR A RECTANGULAR

Programa que muestre la forma rectangular de un numero con forma polar dado

In [18]:
def Polar_Rec(C1):
    print("----------FORMA RECTANGULAR A POLAR-------")
    a = C1[0] * math.cos(C1[1])
    b = C1[0] * math.sin(C1[1])
    print("p = ", C1[0])
    print("Teta = ",C1[1])
    
    print("----------RESULTADO-------")
    print("El resultado en forma rectangular es: (",a,",",b,")" )
    return (a,b)

In [19]:
Polar_Rec([5.39, 0.38])

----------FORMA RECTANGULAR A POLAR-------
p =  5.39
Teta =  0.38
----------RESULTADO-------
El resultado en forma rectangular es: ( 5.0055023857573895 , 1.9992613301359765 )


(5.0055023857573895, 1.9992613301359765)

#### PRUEBA
Se realizo las siguientes pruebas con dos numero complejos

$$ p =  5.39 $$
$$ Teta =  0.38  $$
Y se obtuvo como resultado el siguiente numero
$$ ( 5.0055023857573895 , 1.9992613301359765 ) $$

## CAPITULO 2

### 0. IMPRIMIR VECTORES COMPLEJOS

In [20]:
def imprimir_vec(V1, tam):
    print("Vector complejo:")
    for j in range(tam):
        if V1[j][0]==0:
            print("|  "+str(V1[j][1])+"i    |  ")
        elif V1[j][1]==0:
            print("|   "+str(V1[j][0])+"    |")
        else:
            print("|(",V1[j][0],"+"+str(V1[j][1])+"i )|  ")

### 1.ADICION DE VECTORES COMPLEJOS.

In [21]:
def adi_vectores(V1, V2):
    ## Se define un tamaño y una lista que guarde el resultado componene
    ## a componente
    tam = len(V1)
    Resul = []
    if tam==len(V2):
        for i in range(tam):
            Resul.append(suma_C(V1[i], V2[i]))
        print("----------Adición de vectores complejos-------")
        imprimir_vec(V1, len(V1))
        imprimir_vec(V2, len(V2))
        print("El resultado: ")
        imprimir_vec(Resul, len(Resul))
              
        return Resul

In [22]:
adi_vectores([(8, 3), (-1, -4), (0,-9)],[(8, -3),(2, 5), (3,0)])

----------SUMA DE DOS NÚMEROS-------
Numero complejo : 8 +3i
Numero complejo : 8 -3i
----------RESULTADO-------
Numero complejo : 16
----------SUMA DE DOS NÚMEROS-------
Numero complejo : -1 -4i
Numero complejo : 2 +5i
----------RESULTADO-------
Numero complejo : 1 +1i
----------SUMA DE DOS NÚMEROS-------
Numero complejo : -9i
Numero complejo : 3
----------RESULTADO-------
Numero complejo : 3 -9i
----------Adición de vectores complejos-------
Vector complejo:
|( 8 +3i )|  
|( -1 +-4i )|  
|  -9i    |  
Vector complejo:
|( 8 +-3i )|  
|( 2 +5i )|  
|   3    |
El resultado: 
Vector complejo:
|   16    |
|( 1 +1i )|  
|( 3 +-9i )|  


[(16, 0), (1, 1), (3, -9)]

### 2. MULTIPLICACION DE ESCALAR POR VECTOR COMPLEJO.

In [23]:
def Mul_esc_vect(e, V1):
    Resul = []
    for i in range(len(V1)):
        Resul.append(Mul_C(e,V1[i]))
    
    
    print("----------Multiplicacion de escalar por vector complejo-------")
    print("Escalar")
    imprimir(e)
    imprimir_vec(V1, len(V1))
    print("El resultado: ")
    imprimir_vec(Resul, len(Resul))
    
    return Resul

In [24]:
Mul_esc_vect((-1,1) , [(-2,5),(-1,-1),(2,-9)])

----------MULTIPLICACIÓN DE DOS NÚMEROS-------
Numero complejo : -1 +1i
Numero complejo : -2 +5i
----------RESULTADO-------
Numero complejo : -3 -7i
----------MULTIPLICACIÓN DE DOS NÚMEROS-------
Numero complejo : -1 +1i
Numero complejo : -1 -1i
----------RESULTADO-------
Numero complejo : 2
----------MULTIPLICACIÓN DE DOS NÚMEROS-------
Numero complejo : -1 +1i
Numero complejo : 2 -9i
----------RESULTADO-------
Numero complejo : 7 +11i
----------Multiplicacion de escalar por vector complejo-------
Escalar
Numero complejo : -1 +1i
Vector complejo:
|( -2 +5i )|  
|( -1 +-1i )|  
|( 2 +-9i )|  
El resultado: 
Vector complejo:
|( -3 +-7i )|  
|   2    |
|( 7 +11i )|  


[(-3, -7), (2, 0), (7, 11)]

### 3. INVERSO ADITIVO DE VECTOR COMPLEJO

In [25]:
def inve_vector(V1):
    Resul = Mul_esc_vect((-1,0),V1)
    return Resul
    

In [26]:
inve_vector([(-5, 2),(3, 0),(0, -1)])

----------MULTIPLICACIÓN DE DOS NÚMEROS-------
Numero complejo : -1
Numero complejo : -5 +2i
----------RESULTADO-------
Numero complejo : 5 -2i
----------MULTIPLICACIÓN DE DOS NÚMEROS-------
Numero complejo : -1
Numero complejo : 3
----------RESULTADO-------
Numero complejo : -3
----------MULTIPLICACIÓN DE DOS NÚMEROS-------
Numero complejo : -1
Numero complejo : -1i
----------RESULTADO-------
Numero complejo : 1i
----------Multiplicacion de escalar por vector complejo-------
Escalar
Numero complejo : -1
Vector complejo:
|( -5 +2i )|  
|   3    |
|  -1i    |  
El resultado: 
Vector complejo:
|( 5 +-2i )|  
|   -3    |
|  1i    |  


[(5, -2), (-3, 0), (0, 1)]

### 0.2 IMPRIMIR MATRICES

In [27]:
def imprimir_mat(V1, fila1, columna1):
    print("Matriz compleja:")
    for i in range(fila1):
            for j in range(columna1):
                print("|(",V1[i][j][0],"+"+str(V1[i][j][1])+"i )| ", end="")
            print("")

### 4. SUMA DE MATRICES COMPLEJAS

In [28]:
def Adi_matrices(V1, V2):
    fila1 = len(V1)
    columna1 = len(V1[0])
    fila2 = len(V2)
    columna2 = len(V2[0])
    if fila1==fila2 and columna1==columna2:
        Resul = []
        for i in range(fila1):
            Resul.append([])
            for j in range(columna1):
                Resul[i].append(suma_C(V1[i][j], V2[i][j]))
                
        print("----------Adición de matrices complejos.-------")
        imprimir_mat(V1,fila1, columna1)
        imprimir_mat(V2,fila2, columna2)
        print("El resultado: ")
        imprimir_mat(Resul, fila1, columna1)
        
    
    return Resul
    
    
    

In [29]:
Adi_matrices([[(-8,-3), (-6,-4), (0 ,-4)],[(-1,8), (6,-10),(8,-5)],[(4,0),(8,5),(-7,-9)]],[[(-7, -2),(-4, -2),(7,7)],[(5,9),(0,3),(6,-5)],[(1,5),(-6,-6),(5,8)]])

----------SUMA DE DOS NÚMEROS-------
Numero complejo : -8 -3i
Numero complejo : -7 -2i
----------RESULTADO-------
Numero complejo : -15 -5i
----------SUMA DE DOS NÚMEROS-------
Numero complejo : -6 -4i
Numero complejo : -4 -2i
----------RESULTADO-------
Numero complejo : -10 -6i
----------SUMA DE DOS NÚMEROS-------
Numero complejo : -4i
Numero complejo : 7 +7i
----------RESULTADO-------
Numero complejo : 7 +3i
----------SUMA DE DOS NÚMEROS-------
Numero complejo : -1 +8i
Numero complejo : 5 +9i
----------RESULTADO-------
Numero complejo : 4 +17i
----------SUMA DE DOS NÚMEROS-------
Numero complejo : 6 -10i
Numero complejo : 3i
----------RESULTADO-------
Numero complejo : 6 -7i
----------SUMA DE DOS NÚMEROS-------
Numero complejo : 8 -5i
Numero complejo : 6 -5i
----------RESULTADO-------
Numero complejo : 14 -10i
----------SUMA DE DOS NÚMEROS-------
Numero complejo : 4
Numero complejo : 1 +5i
----------RESULTADO-------
Numero complejo : 5 +5i
----------SUMA DE DOS NÚMEROS-------
Numero 

[[(-15, -5), (-10, -6), (7, 3)],
 [(4, 17), (6, -7), (14, -10)],
 [(5, 5), (2, -1), (-2, -1)]]

### 5. MULTIPLICACION DE ESCALAR POR MATRIZ COMPLEJA

In [30]:
def Mul_esc_matriz(e, V):
    fila1 = len(V)
    columna1 = len(V[0])
    Resul = []
    for i in range(fila1):
        Resul.append([])
        for j in range(columna1):
            Resul[i].append(Mul_C(e,V[i][j]))
    print("----------Multiplicación escalar de matrices complejas-------")
    
    imprimir_mat(V,fila1, columna1)
    print("El resultado: ")
    imprimir_mat(Resul, fila1, columna1)
        
    return Resul
                

In [31]:
Mul_esc_matriz((-2,3), [[(3,-2),(8,-4)],[(4,-10),(-2,-8)]])

----------MULTIPLICACIÓN DE DOS NÚMEROS-------
Numero complejo : -2 +3i
Numero complejo : 3 -2i
----------RESULTADO-------
Numero complejo : 13i
----------MULTIPLICACIÓN DE DOS NÚMEROS-------
Numero complejo : -2 +3i
Numero complejo : 8 -4i
----------RESULTADO-------
Numero complejo : -4 +32i
----------MULTIPLICACIÓN DE DOS NÚMEROS-------
Numero complejo : -2 +3i
Numero complejo : 4 -10i
----------RESULTADO-------
Numero complejo : 22 +32i
----------MULTIPLICACIÓN DE DOS NÚMEROS-------
Numero complejo : -2 +3i
Numero complejo : -2 -8i
----------RESULTADO-------
Numero complejo : 28 +10i
----------Multiplicación escalar de matrices complejas-------
Matriz compleja:
|( 3 +-2i )| |( 8 +-4i )| 
|( 4 +-10i )| |( -2 +-8i )| 
El resultado: 
Matriz compleja:
|( 0 +13i )| |( -4 +32i )| 
|( 22 +32i )| |( 28 +10i )| 


[[(0, 13), (-4, 32)], [(22, 32), (28, 10)]]

### 6. INVERSO ADITIVO DE MATRIZ COMPLEJA

In [32]:
def inv_adi(M1):
    Resul = Mul_esc_matriz((-1,0), M1)
    return Resul

In [33]:
inv_adi([[(7,3),(-1,7)],[(-9,-4),(-7,-9)]])

----------MULTIPLICACIÓN DE DOS NÚMEROS-------
Numero complejo : -1
Numero complejo : 7 +3i
----------RESULTADO-------
Numero complejo : -7 -3i
----------MULTIPLICACIÓN DE DOS NÚMEROS-------
Numero complejo : -1
Numero complejo : -1 +7i
----------RESULTADO-------
Numero complejo : 1 -7i
----------MULTIPLICACIÓN DE DOS NÚMEROS-------
Numero complejo : -1
Numero complejo : -9 -4i
----------RESULTADO-------
Numero complejo : 9 +4i
----------MULTIPLICACIÓN DE DOS NÚMEROS-------
Numero complejo : -1
Numero complejo : -7 -9i
----------RESULTADO-------
Numero complejo : 7 +9i
----------Multiplicación escalar de matrices complejas-------
Matriz compleja:
|( 7 +3i )| |( -1 +7i )| 
|( -9 +-4i )| |( -7 +-9i )| 
El resultado: 
Matriz compleja:
|( -7 +-3i )| |( 1 +-7i )| 
|( 9 +4i )| |( 7 +9i )| 


[[(-7, -3), (1, -7)], [(9, 4), (7, 9)]]

### 7. MATRIZ TRANSPUESTA

In [34]:
def matriz_tran(M1):
    fila1 = len(M1)
    columna1 = len(M1[0])
    Resul = []
    for i in range(columna1):
        aux = []
        for j in range(fila1):
            aux.append(M1[j][i])
        Resul.append(aux)
    print("----------Transpuesta de matriz compleja-------")
    
    imprimir_mat(M1,fila1, columna1)
    print("El resultado: ")
    imprimir_mat(Resul, columna1 , fila1 )        
    return Resul
    

In [35]:
matriz_tran([[(5, 9), (-7, -5), (-1,-4)], [(8,2),(-3,-7), (7, -8)]])

----------Transpuesta de matriz compleja-------
Matriz compleja:
|( 5 +9i )| |( -7 +-5i )| |( -1 +-4i )| 
|( 8 +2i )| |( -3 +-7i )| |( 7 +-8i )| 
El resultado: 
Matriz compleja:
|( 5 +9i )| |( 8 +2i )| 
|( -7 +-5i )| |( -3 +-7i )| 
|( -1 +-4i )| |( 7 +-8i )| 


[[(5, 9), (8, 2)], [(-7, -5), (-3, -7)], [(-1, -4), (7, -8)]]

### 8. CONJUGADA DE UNA MATRIZ COMPLEJA

In [36]:
def Matriz_conju(M1):
    fila1 = len(M1)
    columna1 = len(M1[0])
    Resul = []
    for i in range(fila1):
        Resul.append([])
        for j in range(columna1):
            Resul[i].append(Conju_C(M1[i][j]))
            
    print("----------Conjugada de matriz compleja-------")
    imprimir_mat(M1,fila1, columna1)
    print("El resultado: ")
    imprimir_mat(Resul,fila1 ,columna1)        
    return Resul


In [37]:
Matriz_conju([[(-6, 1), (3, 8)], [(2, -6), (3, 0)]])

----------CONJUGADO DE UN NÚMERO-------
Numero complejo : -6 +1i
----------RESULTADO-------
Numero complejo : -6 -1i
----------CONJUGADO DE UN NÚMERO-------
Numero complejo : 3 +8i
----------RESULTADO-------
Numero complejo : 3 -8i
----------CONJUGADO DE UN NÚMERO-------
Numero complejo : 2 -6i
----------RESULTADO-------
Numero complejo : 2 +6i
----------CONJUGADO DE UN NÚMERO-------
Numero complejo : 3
----------RESULTADO-------
Numero complejo : 3
----------Conjugada de matriz compleja-------
Matriz compleja:
|( -6 +1i )| |( 3 +8i )| 
|( 2 +-6i )| |( 3 +0i )| 
El resultado: 
Matriz compleja:
|( -6 +-1i )| |( 3 +-8i )| 
|( 2 +6i )| |( 3 +0i )| 


[[(-6, -1), (3, -8)], [(2, 6), (3, 0)]]

### 9. MATRIZ ADJUNTA(DAGA)

In [38]:
def Matriz_adj(M1):
    fila1 = len(M1)
    columna1 = len(M1[0])
    Resul = matriz_tran(M1)
    Resul = Matriz_conju(Resul)
    
    print("----------Conjugada de matriz adjunta-------")
    imprimir_mat(M1,fila1, columna1)
    print("El resultado: ")
    imprimir_mat(Resul,columna1, fila1)        
    return Resul

In [39]:
Matriz_adj([[(7, 7), (3, 8),(8,4)], [(5, 0), (8, -6),(-10,-1)]])

----------Transpuesta de matriz compleja-------
Matriz compleja:
|( 7 +7i )| |( 3 +8i )| |( 8 +4i )| 
|( 5 +0i )| |( 8 +-6i )| |( -10 +-1i )| 
El resultado: 
Matriz compleja:
|( 7 +7i )| |( 5 +0i )| 
|( 3 +8i )| |( 8 +-6i )| 
|( 8 +4i )| |( -10 +-1i )| 
----------CONJUGADO DE UN NÚMERO-------
Numero complejo : 7 +7i
----------RESULTADO-------
Numero complejo : 7 -7i
----------CONJUGADO DE UN NÚMERO-------
Numero complejo : 5
----------RESULTADO-------
Numero complejo : 5
----------CONJUGADO DE UN NÚMERO-------
Numero complejo : 3 +8i
----------RESULTADO-------
Numero complejo : 3 -8i
----------CONJUGADO DE UN NÚMERO-------
Numero complejo : 8 -6i
----------RESULTADO-------
Numero complejo : 8 +6i
----------CONJUGADO DE UN NÚMERO-------
Numero complejo : 8 +4i
----------RESULTADO-------
Numero complejo : 8 -4i
----------CONJUGADO DE UN NÚMERO-------
Numero complejo : -10 -1i
----------RESULTADO-------
Numero complejo : -10 +1i
----------Conjugada de matriz compleja-------
Matriz complej

[[(7, -7), (5, 0)], [(3, -8), (8, 6)], [(8, -4), (-10, 1)]]

### 10. PRODUCTO ENTRE MATRICES

In [40]:
def mul_pivote(F1, F2):
    R = Mul_C(F1[0],F2[0])
    suma = R
    for i in range(1,len(F1)):
        R = Mul_C(F1[i],F2[i])
        suma = suma_C(suma,R)
    return suma
       
        

In [41]:
def produc_matriz(M1, M2):
    fila1 = len(M1)
    columna1 = len(M1[0])
    fila2 = len(M2)
    columna2 = len(M2[0])
    Resul = []
    if columna1==fila2:
        transpuesta = matriz_tran(M2)
        for i in range(fila1):
            Resul.append([])
            for j in range(columna2):
                s = mul_pivote(M1[i],transpuesta[j])
                Resul[i].append(s)
        print("----------Conjugada de matriz adjunta-------")
        imprimir_mat(M1,fila1, columna1)
        imprimir_mat(M2,fila2, columna2)
        print("El resultado: ")
        imprimir_mat(Resul,fila1, columna2)        
        return Resul
    else:
        print("ERROR!")
        return 0
    

In [42]:
produc_matriz([[(-6,2),(0,6),(7,2)],[(6,9),(7,7),(-6,-6)],[(5,8),(-6,8),(6,9)]],[[(9,-6),(-3,-4),(5,-2)],[(3,6),(-1,-5),(0,-5)],[(9,9),(8,-4),(-8,-4)]])

----------Transpuesta de matriz compleja-------
Matriz compleja:
|( 9 +-6i )| |( -3 +-4i )| |( 5 +-2i )| 
|( 3 +6i )| |( -1 +-5i )| |( 0 +-5i )| 
|( 9 +9i )| |( 8 +-4i )| |( -8 +-4i )| 
El resultado: 
Matriz compleja:
|( 9 +-6i )| |( 3 +6i )| |( 9 +9i )| 
|( -3 +-4i )| |( -1 +-5i )| |( 8 +-4i )| 
|( 5 +-2i )| |( 0 +-5i )| |( -8 +-4i )| 
----------MULTIPLICACIÓN DE DOS NÚMEROS-------
Numero complejo : -6 +2i
Numero complejo : 9 -6i
----------RESULTADO-------
Numero complejo : -42 +54i
----------MULTIPLICACIÓN DE DOS NÚMEROS-------
Numero complejo : 6i
Numero complejo : 3 +6i
----------RESULTADO-------
Numero complejo : -36 +18i
----------SUMA DE DOS NÚMEROS-------
Numero complejo : -42 +54i
Numero complejo : -36 +18i
----------RESULTADO-------
Numero complejo : -78 +72i
----------MULTIPLICACIÓN DE DOS NÚMEROS-------
Numero complejo : 7 +2i
Numero complejo : 9 +9i
----------RESULTADO-------
Numero complejo : 45 +81i
----------SUMA DE DOS NÚMEROS-------
Numero complejo : -78 +72i
Numero c

[[(-33, 153), (120, 0), (-44, -22)],
 [(87, 0), (-26, -117), (107, 70)],
 [(0, 165), (147, 26), (69, -36)]]

In [43]:
produc_matriz([[(2,1),(3,0),(1,-1)],[(0,0),(0,-2),(7,-3)],[(3,0),(0,0),(1,-2)]],[[(0,-1),(1,0)],[(0,0),(0,1)]])

ERROR!


0

### 11 ACCION DE MATRIZ COMPLEJA SOBRE VECTOR COMPLEJO

In [44]:
def accion(M,V):
    fila1 = len(M)
    columna1 = len(M[0])
    Resul = []
    fila2 = len(V)
    if columna1==fila2:
        for i in range(fila1):
            s = mul_pivote(M[i],V)
            Resul.append(s)
    print("----------Acción de matriz compleja sobre vector complejo-------")
    imprimir_vec(V, fila2)
    imprimir_mat(M,fila1, columna1)
    print("El resultado: ")
    imprimir_vec(Resul, len(Resul))
    return Resul

In [45]:
accion([[(1,-1),(0,-1)],[(0,5),(0,0)]],[(1,1),(2,0)])

----------MULTIPLICACIÓN DE DOS NÚMEROS-------
Numero complejo : 1 -1i
Numero complejo : 1 +1i
----------RESULTADO-------
Numero complejo : 2
----------MULTIPLICACIÓN DE DOS NÚMEROS-------
Numero complejo : -1i
Numero complejo : 2
----------RESULTADO-------
Numero complejo : -2i
----------SUMA DE DOS NÚMEROS-------
Numero complejo : 2
Numero complejo : -2i
----------RESULTADO-------
Numero complejo : 2 -2i
----------MULTIPLICACIÓN DE DOS NÚMEROS-------
Numero complejo : 5i
Numero complejo : 1 +1i
----------RESULTADO-------
Numero complejo : -5 +5i
----------MULTIPLICACIÓN DE DOS NÚMEROS-------
Numero complejo : 0i
Numero complejo : 2
----------RESULTADO-------
Numero complejo : 0i
----------SUMA DE DOS NÚMEROS-------
Numero complejo : -5 +5i
Numero complejo : 0i
----------RESULTADO-------
Numero complejo : -5 +5i
----------Acción de matriz compleja sobre vector complejo-------
Vector complejo:
|( 1 +1i )|  
|   2    |
Matriz compleja:
|( 1 +-1i )| |( 0 +-1i )| 
|( 0 +5i )| |( 0 +0i )| 

[(2, -2), (-5, 5)]

### 12.PRODUCTO INTERNO DE VECTORES COMPLEJOS

In [46]:
def pro_punto(V1,V2):
    fila1 = len(V1)
    fila2 = len(V2)
    S = (0,0)
    if fila1==fila2:
        
        for i in range(fila1):
            S = suma_C(S,Mul_C(Conju_C(V1[i]), V2[i]))
                
        print("----------Producto interno de vectores complejos-------")
        imprimir_vec(V1, len(V1))
        imprimir_vec(V2, len(V2))
        print("El resultado: ")
        imprimir(S)     
        return S
    else:
        return 0
    

In [47]:
pro_punto([(2,-1),(-8,-5),(-2,-6)],[(6,-3),(5,-1),(-6,-2)])

----------CONJUGADO DE UN NÚMERO-------
Numero complejo : 2 -1i
----------RESULTADO-------
Numero complejo : 2 +1i
----------MULTIPLICACIÓN DE DOS NÚMEROS-------
Numero complejo : 2 +1i
Numero complejo : 6 -3i
----------RESULTADO-------
Numero complejo : 15
----------SUMA DE DOS NÚMEROS-------
Numero complejo : 0i
Numero complejo : 15
----------RESULTADO-------
Numero complejo : 15
----------CONJUGADO DE UN NÚMERO-------
Numero complejo : -8 -5i
----------RESULTADO-------
Numero complejo : -8 +5i
----------MULTIPLICACIÓN DE DOS NÚMEROS-------
Numero complejo : -8 +5i
Numero complejo : 5 -1i
----------RESULTADO-------
Numero complejo : -35 +33i
----------SUMA DE DOS NÚMEROS-------
Numero complejo : 15
Numero complejo : -35 +33i
----------RESULTADO-------
Numero complejo : -20 +33i
----------CONJUGADO DE UN NÚMERO-------
Numero complejo : -2 -6i
----------RESULTADO-------
Numero complejo : -2 +6i
----------MULTIPLICACIÓN DE DOS NÚMEROS-------
Numero complejo : -2 +6i
Numero complejo : -6

(4, 1)

### 13. NORMA DE VECTOR COMPLEJO

In [48]:
def norma_vec(V):
    V2 = pro_punto(V,V)
    suma = 0
    for i in range(len(V2)):
        print(V2[i])
        suma += V2[i]
    Resul = math.sqrt(suma)
    print("----------Norma de vectores complejos-------")
    print("Resultado:",int(Resul))
    return Resul
    

In [49]:
norma_vec([(4,5),(3,1),(0,-7)])

----------CONJUGADO DE UN NÚMERO-------
Numero complejo : 4 +5i
----------RESULTADO-------
Numero complejo : 4 -5i
----------MULTIPLICACIÓN DE DOS NÚMEROS-------
Numero complejo : 4 -5i
Numero complejo : 4 +5i
----------RESULTADO-------
Numero complejo : 41
----------SUMA DE DOS NÚMEROS-------
Numero complejo : 0i
Numero complejo : 41
----------RESULTADO-------
Numero complejo : 41
----------CONJUGADO DE UN NÚMERO-------
Numero complejo : 3 +1i
----------RESULTADO-------
Numero complejo : 3 -1i
----------MULTIPLICACIÓN DE DOS NÚMEROS-------
Numero complejo : 3 -1i
Numero complejo : 3 +1i
----------RESULTADO-------
Numero complejo : 10
----------SUMA DE DOS NÚMEROS-------
Numero complejo : 41
Numero complejo : 10
----------RESULTADO-------
Numero complejo : 51
----------CONJUGADO DE UN NÚMERO-------
Numero complejo : -7i
----------RESULTADO-------
Numero complejo : 7i
----------MULTIPLICACIÓN DE DOS NÚMEROS-------
Numero complejo : 7i
Numero complejo : -7i
----------RESULTADO-------
Num

10.0

### 14. DISTANCIA ENTRE VECTORES

In [50]:
def distancia(V1,V2):
    suma = adi_vectores(V1,inve_vector(V2))
    Resul = norma_vec(suma)
    return Resul

In [51]:
distancia([(2,7),(4,-1),(2,-4)],[(7,8),(2,-8),(1,4)])

----------MULTIPLICACIÓN DE DOS NÚMEROS-------
Numero complejo : -1
Numero complejo : 7 +8i
----------RESULTADO-------
Numero complejo : -7 -8i
----------MULTIPLICACIÓN DE DOS NÚMEROS-------
Numero complejo : -1
Numero complejo : 2 -8i
----------RESULTADO-------
Numero complejo : -2 +8i
----------MULTIPLICACIÓN DE DOS NÚMEROS-------
Numero complejo : -1
Numero complejo : 1 +4i
----------RESULTADO-------
Numero complejo : -1 -4i
----------Multiplicacion de escalar por vector complejo-------
Escalar
Numero complejo : -1
Vector complejo:
|( 7 +8i )|  
|( 2 +-8i )|  
|( 1 +4i )|  
El resultado: 
Vector complejo:
|( -7 +-8i )|  
|( -2 +8i )|  
|( -1 +-4i )|  
----------SUMA DE DOS NÚMEROS-------
Numero complejo : 2 +7i
Numero complejo : -7 -8i
----------RESULTADO-------
Numero complejo : -5 -1i
----------SUMA DE DOS NÚMEROS-------
Numero complejo : 4 -1i
Numero complejo : -2 +8i
----------RESULTADO-------
Numero complejo : 2 +7i
----------SUMA DE DOS NÚMEROS-------
Numero complejo : 2 -4i
N

12.0

### 15. Matriz unitaria

In [69]:
def mat_uni(M):
    fila1 = len(M)
    columna1 = len(M[0])
    identidad = []
    for i in range(fila1):
        identidad.append([])
        for j in range(columna1):
            if i==j:
                identidad[i].append(1)
            else:
                identidad[i].append(0)
    daga = Matriz_adj(M)
    Resul = produc_matriz(M,daga)
    
    print("----------¿Es la matriz compleja una matriz unitaria?-------")
    print("El resultado: ")
    print(identidad == Resul)
   
            
        
    

In [70]:
mat_uni([[(0,1),(1,0),(0,0)],[(0,0),(0,1),(1,0)],[(1,0),(0,0),(0,1)]])

----------Transpuesta de matriz compleja-------
Matriz compleja:
|( 0 +1i )| |( 1 +0i )| |( 0 +0i )| 
|( 0 +0i )| |( 0 +1i )| |( 1 +0i )| 
|( 1 +0i )| |( 0 +0i )| |( 0 +1i )| 
El resultado: 
Matriz compleja:
|( 0 +1i )| |( 0 +0i )| |( 1 +0i )| 
|( 1 +0i )| |( 0 +1i )| |( 0 +0i )| 
|( 0 +0i )| |( 1 +0i )| |( 0 +1i )| 
----------CONJUGADO DE UN NÚMERO-------
Numero complejo : 1i
----------RESULTADO-------
Numero complejo : -1i
----------CONJUGADO DE UN NÚMERO-------
Numero complejo : 0i
----------RESULTADO-------
Numero complejo : 0i
----------CONJUGADO DE UN NÚMERO-------
Numero complejo : 1
----------RESULTADO-------
Numero complejo : 1
----------CONJUGADO DE UN NÚMERO-------
Numero complejo : 1
----------RESULTADO-------
Numero complejo : 1
----------CONJUGADO DE UN NÚMERO-------
Numero complejo : 1i
----------RESULTADO-------
Numero complejo : -1i
----------CONJUGADO DE UN NÚMERO-------
Numero complejo : 0i
----------RESULTADO-------
Numero complejo : 0i
----------CONJUGADO DE UN NÚM

### 16. MATRIZ HERMITIANA

In [ ]:
def mat_her(M):
    daga = Matriz_adj(M)
    
    print("----------¿Es la matriz compleja una matriz hermitiana?-------")
    print("El resultado: ")
    print(M == daga)

In [ ]:
mat_her([[(3,0),(2,-1),(0,-3)],[(2,1),(0,0),(1,-1)],[(0,3),(1,1),(0,0)]])

### Producto Tesorial entre matrices complejas

In [ ]:
def prod_tensor(M1,M2):
    fila1 = len(M1)
    columna1 = len(M1[0])
    fila2 = len(M2)
    columna2 = len(M2[0])
    Resul = []
    for i in range(fila1):
        for j in range(columna1):
            Res1 = Mul_esc_matriz(M1[i][j],M2)
            